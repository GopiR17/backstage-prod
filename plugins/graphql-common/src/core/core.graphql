directive @field(
  at: _DirectiveArgument_!
  default: _DirectiveArgument_
) on FIELD_DEFINITION
directive @inherit(
  interface: String
  generatedTypeName: String
  when: _DirectiveArgument_
  is: _DirectiveArgument_
) on INTERFACE

scalar _DirectiveArgument_

interface Node {
  id: ID!
}

interface Connection @inherit(generatedTypeName: "NodeConnection") {
  pageInfo: PageInfo!
  edges: [Edge!]!
  count: Int
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
  endCursor: String
}

interface Edge @inherit(generatedTypeName: "NodeEdge") {
  cursor: String!
  node: Node!
}

type Query {
  node(id: ID!): Node
}
