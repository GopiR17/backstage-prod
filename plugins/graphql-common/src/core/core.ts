/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { resolvePackagePath } from '@backstage/backend-common';
import { loadFilesSync } from '@graphql-tools/load-files';
import { createModule } from 'graphql-modules';
import type { ResolverContext } from '../types';

export const Core = createModule({
  id: 'core',
  typeDefs: loadFilesSync(
    resolvePackagePath(
      '@backstage/plugin-graphql-common',
      'src/core/core.graphql',
    ),
  ),
  resolvers: {
    Node: {
      id: async (
        { id }: { id: string },
        _: never,
        { loader }: ResolverContext,
      ): Promise<string | null> => {
        const entity = await loader.load(id);
        if (!entity) return null;
        return id;
      },
    },
    Query: {
      node: (_: any, { id }: { id: string }): { id: string } => ({ id }),
      entity: (
        _: any,
        {
          name,
          kind,
          namespace = 'default',
        }: { name: string; kind: string; namespace: string },
        { refToId }: ResolverContext,
      ): { id: string } => ({ id: refToId({ name, kind, namespace }) }),
    },
  },
});
