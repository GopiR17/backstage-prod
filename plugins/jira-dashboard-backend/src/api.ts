/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Config } from '@backstage/config';
import fetch from 'node-fetch';
import {
  Project,
  Filter,
  Issue,
  JiraIssue,
  NewJiraIssueResponse,
} from './types';
import { JIRA_BASE_URL_CONFIG_PATH, JIRA_TOKEN_CONFIG_PATH } from './constants';

export const getProjectInfo = async (
  projectKey: string,
  config: Config,
): Promise<Project> => {
  const response = await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}project/${projectKey}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  );
  if (response.status !== 200) {
    throw Error(`${response.status}`);
  }
  return response.json();
};

export const getFilterById = async (
  id: string,
  config: Config,
): Promise<Filter> => {
  const response = await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}filter/${id}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  );
  if (response.status !== 200) {
    throw Error(`${response.status}`);
  }
  const jsonResponse = await response.json();
  return { name: jsonResponse.name, query: jsonResponse.jql } as Filter;
};

export const getIssuesByFilter = async (
  projectKey: string,
  query: string,
  config: Config,
): Promise<Issue[]> => {
  const response = await fetch(
    `${config.getString(
      JIRA_BASE_URL_CONFIG_PATH,
    )}search?jql=project=${projectKey} AND ${query}`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  ).then(resp => resp.json());
  return response.issues;
};

export const getIssuesByComponent = async (
  projectKey: string,
  componentKey: string,
  config: Config,
): Promise<Issue[]> => {
  const response = await fetch(
    `${config.getString(
      JIRA_BASE_URL_CONFIG_PATH,
    )}search?jql=project=${projectKey} AND component = "${componentKey}"`,
    {
      method: 'GET',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
      },
    },
  ).then(resp => resp.json());
  return response.issues;
};

export async function getProjectAvatar(url: string, config: Config) {
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
    },
  });
  return response;
}

export async function postNewIssue(
  jiraIssue: JiraIssue,
  config: Config,
): Promise<NewJiraIssueResponse> {
  const body = JSON.stringify({ fields: jiraIssue });
  return await fetch(`${config.getString(JIRA_BASE_URL_CONFIG_PATH)}issue`, {
    method: 'POST',
    headers: {
      Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body,
  }).then(response => response.json());
}

export async function addIssueWatcher(
  jira_id: string,
  user: string,
  config: Config,
): Promise<void> {
  await fetch(
    `${config.getString(JIRA_BASE_URL_CONFIG_PATH)}issue/${jira_id}/watchers`,
    {
      method: 'POST',
      headers: {
        Authorization: `${config.getString(JIRA_TOKEN_CONFIG_PATH)}`,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: `"${user}"`,
    },
  );
}
