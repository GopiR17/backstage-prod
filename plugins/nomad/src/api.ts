/*
 * Copyright 2020 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Config } from '@backstage/config';
import { createApiRef } from '@backstage/core-plugin-api';

/** @public */
export const nomadApiRef = createApiRef<NomadApi>({
  id: 'plugin.nomad.service',
});

/** @public */
export type NomadApi = {
  /**
   * listAllocations is for listing all allocations matching some part of 'prefix'.
   *
   * See: https://developer.hashicorp.com/nomad/api-docs/allocations#list-allocations
   */
  listAllocations: (
    options: ListAllocationsRequest,
  ) => Promise<ListAllocationsResponse>;
};

/** @public */
export interface ListAllocationsRequest {
  namespace?: string;
  filter?: string;
}

/** @public */
export interface ListAllocationsResponse {
  allocations: Allocation[];
}

/** @public */
export interface Allocation {
  id: string;
  clientStatus: string;
  createTime: number;
  namespace: string;
  name: string;
  modifyTime: number;
  nodeID: string;
  taskGroup: string;
}

/** @public */
export interface DeploymentStatus {
  healthy: boolean;
  timestamp: string;
}

/** @public */
export class NomadHttpApi implements NomadApi {
  static fromConfig(config: Config) {
    return new NomadHttpApi(
      config.getOptionalString('nomad.nomadAddr'),
      config.getOptionalString('nomad.nomadRegion'),
      config.getOptionalString('nomad.nomadNamespace'),
      config.getOptionalString('nomad.nomadToken'),
    );
  }

  constructor(
    private nomadAddr?: string,
    private nomadRegion?: string,
    private nomadNamespace?: string,
    private nomadToken?: string,
  ) {}

  async listAllocations(
    options: ListAllocationsRequest,
  ): Promise<ListAllocationsResponse> {
    return Promise.resolve({
      allocations: [],
    });
  }
}
