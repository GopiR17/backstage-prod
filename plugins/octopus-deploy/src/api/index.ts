/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  createApiRef,
  DiscoveryApi,
  IdentityApi,
} from '@backstage/core-plugin-api';

export type OctopusProgression = {
  Environments: OctopusEnvironment[];
  Releases: OctopusReleaseProgression[];
};

export type OctopusEnvironment = {
  Id: string;
  Name: string;
};

export type OctopusReleaseProgression = {
  Release: OctopusRelease;
  Deployments: { [key: string]: OctopusDeployment[] };
};

export type OctopusRelease = {
  Id: string;
  Version: string;
};

export type OctopusDeployment = {
  State: string;
};

export const octopusDeployApiRef = createApiRef<OctopusDeployApi>({
  id: 'plugin.octopusdeploy.service',
});

const DEFAULT_PROXY_PATH_BASE = '/octopus-deploy';

type Options = {
  discoveryApi: DiscoveryApi;
  identityApi: IdentityApi;
  /**
   * Path to use for requests via the proxy, defaults to /octopus-deploy
   */
  proxyPathBase?: string;
};

export interface OctopusDeployApi {
  getReleaseProgression(
    projectId: string,
    releaseHistoryCount: number,
  ): Promise<OctopusProgression>;
}

export class OctopusDeployClient implements OctopusDeployApi {
  private readonly discoveryApi: DiscoveryApi;
  private readonly identityApi: IdentityApi;
  private readonly proxyPathBase: string;

  constructor(options: Options) {
    this.discoveryApi = options.discoveryApi;
    this.identityApi = options.identityApi;
    this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
  }

  async getReleaseProgression(
    projectId: string,
    releaseHistoryCount: number,
  ): Promise<OctopusProgression> {
    const url = await this.getApiUrl(projectId, releaseHistoryCount);

    const { token: idToken } = await this.identityApi.getCredentials();
    const response = await fetch(url, {
      headers: idToken ? { Authorization: `Bearer ${idToken}` } : {},
    });

    let responseJson;

    try {
      responseJson = await response.json();
    } catch (e) {
      responseJson = { releases: [] };
    }

    if (response.status !== 200) {
      throw new Error(
        `Error communicating with Octopus Deploy: ${
          responseJson?.error?.title || response.statusText
        }`,
      );
    }

    return responseJson;
  }

  private async getApiUrl(projectId: string, releaseHistoryCount: number) {
    const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
    return `${proxyUrl}${this.proxyPathBase}/projects/${projectId}/progression?releaseHistoryCount=${releaseHistoryCount}`;
  }
}
