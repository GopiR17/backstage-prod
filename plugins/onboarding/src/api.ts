/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  createApiRef,
  DiscoveryApi,
  FetchApi,
  IdentityApi,
} from '@backstage/core-plugin-api';
import { CatalogApi } from '@backstage/catalog-client';

export const onboardingApiRef = createApiRef<OnboardingApi>({
  id: 'onboarding',
});

export interface OnboardingApi {
  getChecklist(groups: string, roles: string): Promise<any>;
  updateChecklistStatus(body: any): Promise<any>;
  getUserInfo(ref: string): Promise<any>;
}

export class OnboardingClient implements OnboardingApi {
  private readonly discoveryApi: DiscoveryApi;
  private readonly fetchApi: FetchApi;
  private readonly catalogApi: CatalogApi;
  constructor(options: {
    identityApi: IdentityApi;
    discoveryApi: DiscoveryApi;
    fetchApi: FetchApi;
    catalogApi: CatalogApi;
  }) {
    this.discoveryApi = options.discoveryApi;
    this.fetchApi = options.fetchApi;
    this.catalogApi = options.catalogApi;
  }

  async getUserInfo(ref: string): Promise<any> {
    return await this.catalogApi.getEntityByRef({
      kind: 'user',
      namespace: 'default',
      name: ref,
    });
  }

  async getChecklist(groups: string, roles: string): Promise<any> {
    const baseUrl = await this.discoveryApi.getBaseUrl('onboarding');
    return await this.fetchApi.fetch(
      `${baseUrl}/checklists?groups=${groups}&roles=${roles}`,
      {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      },
    );
  }
  async updateChecklistStatus(body: any): Promise<any> {
    const baseUrl = await this.discoveryApi.getBaseUrl('onboarding');
    return await this.fetchApi.fetch(`${baseUrl}/update-status`, {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
}
