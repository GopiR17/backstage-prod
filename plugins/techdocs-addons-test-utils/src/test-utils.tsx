/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, Fragment } from 'react';

// Shadow DOM support for the simple and complete DOM testing utilities
// https://github.com/testing-library/dom-testing-library/issues/742#issuecomment-674987855
import { screen } from 'testing-library__dom';
import { renderToStaticMarkup } from 'react-dom/server';
import { Route } from 'react-router-dom';
import { act, render } from '@testing-library/react';

import { wrapInTestApp, TestApiProvider } from '@backstage/test-utils';
import { FlatRoutes } from '@backstage/core-app-api';
import { ApiRef } from '@backstage/core-plugin-api';

import {
  TechDocsAddons,
  techdocsApiRef,
  TechDocsEntityMetadata,
  TechDocsMetadata,
  techdocsStorageApiRef,
} from '@backstage/plugin-techdocs-react';
import { TechDocsReaderPage, techdocsPlugin } from '@backstage/plugin-techdocs';
import { catalogPlugin } from '@backstage/plugin-catalog';
import { searchApiRef } from '@backstage/plugin-search-react';
import { scmIntegrationsApiRef } from '@backstage/integration-react';

const techdocsApi = {
  getTechDocsMetadata: jest.fn(),
  getEntityMetadata: jest.fn(),
};

const techdocsStorageApi = {
  getApiOrigin: jest.fn(),
  getEntityDocs: jest.fn(),
  syncEntityDocs: jest.fn(),
};

const searchApi = {
  query: jest.fn().mockResolvedValue({ results: [] }),
};

const scmIntegrationsApi = {
  fromConfig: jest.fn().mockReturnValue({}),
};

/** @ignore */
type TechDocsAddonTesterTestApiPair<TApi> = TApi extends infer TImpl
  ? readonly [ApiRef<TApi>, Partial<TImpl>]
  : never;

/** @ignore */
type TechdocsAddonTesterApis<T> = TechDocsAddonTesterTestApiPair<T>[];

type TechDocsAddonTesterOptions = {
  dom: ReactElement;
  entity: Partial<TechDocsEntityMetadata>;
  metadata: Partial<TechDocsMetadata>;
  componentId: string;
  apis: TechdocsAddonTesterApis<any>;
  path: string;
};

const defaultOptions: TechDocsAddonTesterOptions = {
  dom: <></>,
  entity: {},
  metadata: {},
  componentId: 'docs',
  apis: [],
  path: '',
};

const defaultMetadata = {
  site_name: 'Tech Docs',
  site_description: 'Tech Docs',
};

const defaultEntity = {
  kind: 'Component',
  metadata: { namespace: 'default', name: 'docs' },
};

const defaultDom = (
  <html lang="en">
    <head />
    <body>
      <div data-md-component="container">
        <div data-md-component="navigation" />
        <div data-md-component="toc" />
        <div data-md-component="main" />
      </div>
    </body>
  </html>
);

/**
 * @public
 */

export class TechDocsAddonTester {
  private options: TechDocsAddonTesterOptions = defaultOptions;
  private addons: ReactElement[];

  static buildAddonsInTechDocs(addons: ReactElement[]) {
    return new TechDocsAddonTester(addons);
  }

  private constructor(addons: ReactElement[]) {
    this.addons = addons;
  }

  withApis<T>(apis: TechdocsAddonTesterApis<T>) {
    const refs = apis.map(([ref]) => ref);
    this.options.apis = this.options.apis
      .filter(([ref]) => !refs.includes(ref))
      .concat(apis);
    return this;
  }

  withDom(dom: ReactElement) {
    this.options.dom = dom;
    return this;
  }

  withMetadata(metadata: Partial<TechDocsMetadata>) {
    this.options.metadata = metadata;
    return this;
  }

  withEntity(entity: Partial<TechDocsEntityMetadata>) {
    this.options.entity = entity;
    return this;
  }

  atPath(path: string) {
    this.options.path = path;
    return this;
  }

  build() {
    const apis: TechdocsAddonTesterApis<any> = [
      [techdocsApiRef, techdocsApi],
      [techdocsStorageApiRef, techdocsStorageApi],
      [searchApiRef, searchApi],
      [scmIntegrationsApiRef, scmIntegrationsApi],
      ...this.options.apis,
    ];

    const entityName = {
      namespace:
        this.options.entity?.metadata?.namespace ||
        defaultEntity.metadata.namespace,
      kind: this.options.entity?.kind || defaultEntity.kind,
      name: this.options.entity?.metadata?.name || defaultEntity.metadata.name,
    };

    techdocsApi.getTechDocsMetadata.mockReturnValue(
      this.options.metadata || { ...defaultMetadata },
    );
    techdocsApi.getEntityMetadata.mockResolvedValue(
      this.options.entity || { ...defaultEntity },
    );

    techdocsStorageApi.syncEntityDocs.mockResolvedValue('cached');
    techdocsStorageApi.getApiOrigin.mockResolvedValue(
      'https://backstage.example.com/api/techdocs',
    );

    techdocsStorageApi.getEntityDocs.mockResolvedValue(
      renderToStaticMarkup(this.options.dom || defaultDom),
    );

    const TechDocsAddonsPage = () => {
      return (
        <TestApiProvider apis={apis}>
          <FlatRoutes>
            <Route
              path="/docs/:namespace/:kind/:name/*"
              element={<TechDocsReaderPage />}
            >
              <TechDocsAddons>
                {this.addons.map((addon, index) => (
                  <Fragment key={index}>{addon}</Fragment>
                ))}
              </TechDocsAddons>
            </Route>
          </FlatRoutes>
        </TestApiProvider>
      );
    };

    return wrapInTestApp(<TechDocsAddonsPage />, {
      routeEntries: [
        `/docs/${entityName.namespace}/${entityName.kind}/${entityName.name}/${this.options.path}`,
      ],
      mountedRoutes: {
        '/docs': techdocsPlugin.routes.root,
        '/docs/:namespace/:kind/:name/*': techdocsPlugin.routes.docRoot,
        '/catalog/:namespace/:kind/:name': catalogPlugin.routes.catalogEntity,
      },
    });
  }

  // Components using useEffect to perform an asynchronous action (such as fetch) must be rendered within an async
  // act call to properly get the final state, even with mocked responses. This utility method makes the signature a bit
  // cleaner, since act doesn't return the result of the evaluated function.
  // https://github.com/testing-library/react-testing-library/issues/281
  // https://github.com/facebook/react/pull/14853
  async renderWithEffects(): Promise<
    typeof screen & { shadowRoot: ShadowRoot | null }
  > {
    await act(async () => {
      render(this.build());
    });

    const shadowHost = screen.getByTestId('techdocs-native-shadowroot');

    return {
      ...screen,
      shadowRoot: shadowHost?.shadowRoot || null,
    };
  }
}

export default TechDocsAddonTester.buildAddonsInTechDocs;
