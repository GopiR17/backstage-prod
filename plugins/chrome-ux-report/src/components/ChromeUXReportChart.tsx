/*
 * Copyright 2021 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { useApi, Progress } from '@backstage/core';
import { chromeuxReportApiRef } from '../api';
import { useAsync } from 'react-use';
import Alert from '@material-ui/lab/Alert';
import { Chart } from 'react-google-charts';
import { getPeriod, getRoundOfMetric } from '../utils';

export const ChromeUXReportChart = ({ origin }: { origin: string }) => {
  const chromeUXReportApi = useApi(chromeuxReportApiRef);

  const { value, loading, error } = useAsync(async (): Promise<any> => {
    const response = await chromeUXReportApi.getChromeUXMetrics(origin);
    return response.metrics;
  }, []);

  if (loading) {
    return <Progress />;
  } else if (error) {
    return <Alert severity="error">{error.message}</Alert>;
  }

  const {
    fast_fp,
    avg_fp,
    slow_fp,
    fast_fcp,
    avg_fcp,
    slow_fcp,
    fast_dcl,
    avg_dcl,
    slow_dcl,
    fast_ol,
    avg_ol,
    slow_ol,
    fast_fid,
    avg_fid,
    slow_fid,
    fast_ttfb,
    avg_ttfb,
    slow_ttfb,
    fast_lcp,
    avg_lcp,
    slow_lcp,
  } = value;

  return (
    <div>
      <div
        className="period"
        style={{
          display: 'flex',
          position: 'absolute',
          zIndex: 99,
          alignItems: 'center',
          justifyContent: 'center',
          width: '95%',
          marginTop: '30px',
          fontSize: '25px',
          fontWeight: 'bold',
          color: 'grey',
        }}
      >
        {getPeriod()}
      </div>
      <Chart
        width="100%"
        height="70vh"
        chartType="BarChart"
        loader={<Progress />}
        data={[
          ['METRICS', 'FAST', 'AVERAGE', 'SLOW'],
          [
            'First Contentful Paint',
            getRoundOfMetric(fast_fcp),
            getRoundOfMetric(avg_fcp),
            getRoundOfMetric(slow_fcp),
          ],
          [
            'Largest Contentful Paint',
            getRoundOfMetric(fast_lcp),
            getRoundOfMetric(avg_lcp),
            getRoundOfMetric(slow_lcp),
          ],
          [
            'Dom Content Loaded',
            getRoundOfMetric(fast_dcl),
            getRoundOfMetric(avg_dcl),
            getRoundOfMetric(slow_dcl),
          ],
          [
            'First Input Delay',
            getRoundOfMetric(fast_fid),
            getRoundOfMetric(avg_fid),
            getRoundOfMetric(slow_fid),
          ],
          [
            'Onload',
            getRoundOfMetric(fast_ol),
            getRoundOfMetric(avg_ol),
            getRoundOfMetric(slow_ol),
          ],
          [
            'First Paint',
            getRoundOfMetric(fast_fp),
            getRoundOfMetric(avg_fp),
            getRoundOfMetric(slow_fp),
          ],
          [
            'Time To First Byte',
            getRoundOfMetric(fast_ttfb),
            getRoundOfMetric(avg_ttfb),
            getRoundOfMetric(slow_ttfb),
          ],
        ]}
        options={{
          isStacked: true,
          hAxis: {
            title: 'PERCENTAGE',
            titleTextStyle: {
              color: 'grey',
              bold: true,
            },
            textStyle: {
              color: '#CCCCCC',
              bold: true,
            },
            viewWindowMode: 'explicit',
            viewWindow: {
              min: 0,
              max: 100,
            },
          },
          animation: {
            startup: true,
            easing: 'linear',
            duration: 500,
          },
          vAxis: {
            title: 'METRICS',
            titleTextStyle: {
              color: 'grey',
              bold: true,
            },
            textStyle: {
              color: '#CCCCCC',
              bold: true,
            },
          },
          legend: {
            textStyle: {
              color: 'grey',
              bold: true,
            },
          },
          backgroundColor: '#424242',
          colors: ['#5cb85c', '#f0ad4e', '#d9534f'],
          chartArea: {
            left: 300,
            top: 130,
          },
        }}
        rootProps={{ 'data-testid': '3' }}
      />
    </div>
  );
};
