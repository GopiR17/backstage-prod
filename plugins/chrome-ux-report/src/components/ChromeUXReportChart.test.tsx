/*
 * Copyright 2021 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { ChromeUXReportChart } from './ChromeUXReportChart';
import { wrapInTestApp } from '@backstage/test-utils';
import { render, waitFor } from '@testing-library/react';
import { ApiProvider, ApiRegistry } from '@backstage/core';
import { ChromeUXReportApi, chromeuxReportApiRef } from '../api';

describe('ChromeUXReportChart', () => {
  const chromeUXReportApi: Partial<ChromeUXReportApi> = {
    getChromeUXMetrics: async () => ({
      metrics: {
        origin_id: 1,
        period_id: 1,
        fast_fp: 0.9,
        avg_fp: 0.08,
        slow_fp: 0.02,
        fast_fcp: 0.9,
        avg_fcp: 0.08,
        slow_fcp: 0.02,
        fast_dcl: 0.9,
        avg_dcl: 0.08,
        slow_dcl: 0.02,
        fast_ol: 0.9,
        avg_ol: 0.08,
        slow_ol: 0.02,
        fast_fid: 0.9,
        avg_fid: 0.08,
        slow_fid: 0.02,
        fast_ttfb: 0.9,
        avg_ttfb: 0.08,
        slow_ttfb: 0.02,
        fast_lcp: 0.9,
        avg_lcp: 0.08,
        slow_lcp: 0.02,
      },
    }),
  };

  const renderWrapped = (children: React.ReactNode) =>
    render(
      wrapInTestApp(
        <ApiProvider
          apis={ApiRegistry.from([[chromeuxReportApiRef, chromeUXReportApi]])}
        >
          {children}
        </ApiProvider>,
      ),
    );

  it('render snapshot', async () => {
    const origin = 'Backstage';
    const { asFragment } = renderWrapped(
      <ChromeUXReportChart origin={origin} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('render successfully', async () => {
    const origin = 'Backstage';
    const { container } = renderWrapped(
      <ChromeUXReportChart origin={origin} />,
    );
    const period = container.getElementsByClassName('period');

    await waitFor(async () => {
      expect(period).toHaveLength(1);
      expect(period).not.toBeNull();
    });
  });
});
