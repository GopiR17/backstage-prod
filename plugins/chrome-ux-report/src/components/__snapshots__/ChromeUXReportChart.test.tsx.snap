// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeUXReportChart render snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: none;"
  />
</DocumentFragment>
`;
